<p><code translate="no">finally</code> ブロックから<code translate="no">return</code>,<code translate="no">break</code>,<code translate="no">throw</code> などを使用すると、<code translate="no">try</code> または<code translate="no">catch</code> ブロックで投げられた未処理の<code translate="no">Throwable</code> の伝搬が抑制されます。</p>
<p>このルールでは、ジャンプ文（<code translate="no">break</code>,<code translate="no">continue</code>,<code translate="no">return</code>,<code translate="no">throw</code>,<code translate="no">goto</code> ）によって、制御フローが<code translate="no">finally</code> ブロックから離れることになる場合に問題が発生します。</p>
<h2>コンプライアンス違反のコード例</h2>
<pre translate="no">
public static void main(String[] args) {
  try {
    doSomethingWhichThrowsException();
    System.out.println("OK");   // incorrect "OK" message is printed
  } catch (RuntimeException e) {
    System.out.println("ERROR");  // this message is not shown
  }
}

public static void doSomethingWhichThrowsException() {
  try {
    throw new RuntimeException();
  } finally {
    for (int i = 0; i &lt; 10; i ++) {
      //...
      if (q == i) {
        break; // ignored
      }
    }

    /* ... */
    return;      // Noncompliant - prevents the RuntimeException from being propagated
  }
}
</pre>
<h2>コンプライアンス・ソリューション</h2>
<pre translate="no">
public static void main(String[] args) {
  try {
    doSomethingWhichThrowsException();
    System.out.println("OK");
  } catch (RuntimeException e) {
    System.out.println("ERROR");  // "ERROR" is printed as expected
  }
}

public static void doSomethingWhichThrowsException() {
  try {
    throw new RuntimeException();
  } finally {
    for (int i = 0; i &lt; 10; i ++) {
      //...
      if (q == i) {
        break; // ignored
      }
    }

    /* ... */
  }
}
</pre>
<h2>見る</h2>
<ul>
  <li><a href="https://cwe.mitre.org/data/definitions/584">MITRE、CWE-584</a>- 最後にブロックの内側を返す。 </li>
  <li><a href="https://wiki.sei.cmu.edu/confluence/x/BTdGBQ">CERT、ERR04-J。</a>- 最終ブロックから急に完了しないこと </li>
</ul>

